frontend/catalyst/jax_primitives.py

```diff
@@ -132,6 +148,15 @@ def __init__(self, num_qubits=None, primitive=None):
         self.num_qubits = num_qubits
         self.primitive = primitive
 
+    def __eq__(self, other):
+        if not isinstance(other, AbstractObs):
+            return False
+
+        return self.num_qubits == other.num_qubits and self.primitive == other.primitive
+
+    def __hash__(self):
+        return hash(self.primitive) + self.num_qubits
```

Comment:
Is the addition to the hash intended? Would it be better to use tuples for combining?
```suggestion
        return hash(("ConcreteQreg",self.primitive, self.num_qubits))
```

Reply:
Yes the hash is needed because the class needs to be useable as a key. Without either `__eq__` or `__hash__` the class is useable as a key by default, but since I added `__eq__` an explicit `__hash__` was needed as well.

I believe the only requirement is that given `A == B` then we must have `hash(A) == hash(B)`, which this implementation should satisfy. I don't see the need to hash the tuple, do you have a reason?

Reply:
I think that using tuples for combining hashes is a more safe default solution (same [here](https://stackoverflow.com/a/2909119)). Other methods might have non-obvious consequences for stability in general. But this use-case is not a critical one (and I am not an expert in hashing) so I am only suggesting.

Reply:
I included `__class__.__name__` to distinguish AbstractObs from the ConcreteObs below.

Reply:
Again I don't think it's relevant at all, and hash functions should be kept as short as possible (in execution time), at least when talking about indexing hash maps and not other applications.

Reply:
Sure, I do not have other arguments regarding this.